### DISCUSSION TITLE
The UC Detection Contest

### DISCUSSION THREAD

**Discussion by HMaker** [ultrafunkamsterdam/undetected-chromedriver#1086] [Ideas]
I think the undetected-chromedriver community should also work on detecting UC flaws, this makes the development more active and antecipates a lot of issues.

Post on this thread new ways you found to detect UC, please include a reproducible example with a Python script for the browser automation and a HTML file which simulates an antibot trying to detect UC.

I made [Selenium Detector](https://github.com/HMaker/HMaker.github.io/tree/master/selenium-detector) which can detect latest UC version.

**Comment by mmbertasi** [comment]
This tool was super super useful for me in the past and I meant to say thank you!  

That being said I have a question:

The only test my selenium is failing is "devtools-console
Either Devtools Console is open or CDP Runtime Domain is enabled".  Is there anyway to avoid this?

Thanks again!

**Comment by donggoing** [comment]
Is there any way to bypass "Either Devtools Console is open or CDP Runtime Domain is enabled"?

**Reply by mmbertasi** [reply]
Yea that's the question; think this is how some sites like b365 are detecting

**Reply by ovciina** [reply]
> Yea that's the question; think this is how some sites like b365 are detecting

Can you elaborate what you mean? I dont really understan the CDP. I will google it and read a bit more but are you sure it's the way Bet365 is detecting? Its been more than one week and still not working to open their site with UC, it has worked for 2-3 years now atleast..

**Reply by mmbertasi** [reply]
As far as I understand it, selenium uses "CDP" aka Chrome Devtools Protocol to interact with the Chrome browser. Developer tools is not open by default in a regular Chrome browser but the CDP framework exposes it when you go the page in Selenium. 
I could be not entirely on point here but that's my understanding so far.

**Reply by ovciina** [reply]
Ah, I understand and just tried manually without chromedriver. You are correct, atleast when devtools are open the site won't load properly.. Damn, they are smart. If Selenium now uses CDP to work we can't really bypass it..

**Reply by mmbertasi** [reply]
I think there probably is a way but it will take someone who understands UC at a more fundamental level...

**Comment by ttlns** [comment]
Found 2 workarounds for the CDP CDP Runtime Domain so far:
1. patch by using javascript (but possiility to detect the patches)
2. disably runtime domain (but causes a lot of bugs, => move to Chrome-Extension, aka Selenium-IDE?)

Question:
- Do you see a possibility to patch the runtime in Chromium directly? So basically somewhere at [/inspector/v8](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/inspector/v8-runtime-agent-impl.cc;l=992)
- If yes, can you give me a starting point to that? Like what would need to be changed in which files? Never worked with C++

**Reply by ramsestom** [reply]
Could you point me to the javascript code you use to evade CDP Runtime Domain detection?
Thanks

**Reply by flen201** [reply]
any examples for the first patching method in js??
